File: TypeScript/src/in_penalty_box.ts
import { Action } from "./action";
import { Base } from "./base";
import { Categories } from "./categories";
import { CurrentPlayer } from "./current_player";
import { PenaltyBox } from "./penalty_box";
import { Places } from "./places";
import { PlayerNames } from "./player_names";
import { Purses } from "./purses";
import { QuestionsByCategory } from "./questions_by_category";

export class InPenaltyBox extends Base implements Action {

    public constructor(
        currentPlayer: CurrentPlayer,
        names: PlayerNames,
        places: Places,
        purses: Purses,
        penaltyBox: PenaltyBox,
        categories: Categories,
        questions: QuestionsByCategory,
        isGettingOutOfPenaltyBox: boolean[]) {
        super(currentPlayer,
            names,
            places,
            purses,
            penaltyBox,
            categories,
            questions,
            isGettingOutOfPenaltyBox);
    }

    public execute(): boolean {
        if (this.isGettingOutOfPenaltyBox[0]) {
            this.println();
            this.purses.inc();

            const winner: boolean = this.didPlayerWin();
            this.currentPlayer.inc();

            return winner;
        } else {
            this.currentPlayer.inc();
            return true;
        }
    }

    private println(): void {
        console.log("Answer was correct!!!!");
    }

    private didPlayerWin(): boolean {
        return !(this.purses.current() === 6);
    }

}

File: Java/src/main/java/com/adaptionsoft/games/uglytrivia/InPenaltyBox.java
package com.adaptionsoft.games.uglytrivia;

public class InPenaltyBox extends Base implements Action {

    public InPenaltyBox(CurrentPlayer currentPlayer,
            PlayerNames names,
            Places places,
            Purses purses,
            PenaltyBox penaltyBox,
            Categories categories,
            QuestionsByCategory questions,
            boolean[] isGettingOutOfPenaltyBox) {
        super(currentPlayer,
                names,
                places,
                purses,
                penaltyBox,
                categories,
                questions,
                isGettingOutOfPenaltyBox);
    }

    @Override
    public boolean execute() {
        if (isGettingOutOfPenaltyBox[0]) {
            println();
            purses.inc();

            boolean winner = didPlayerWin();
            currentPlayer.inc();

            return winner;
        } else {
            currentPlayer.inc();
            return true;
        }
    }

    private void println() {
        System.out.println("Answer was correct!!!!");
    }

    private boolean didPlayerWin() {
        return !(purses.current() == 6);
    }

}
